lms/djangoapps/instructor_task/tasks_helper/misc.py

```diff
@@ -340,3 +349,227 @@ def upload_ora2_data(
+        └── downloads.csv
+    ```
+    """
+
+    storage = DefaultStorage()
+
+    with ZipFile(file, 'w') as zip_file:
+        for block_id, submissions in submission_files_data.items():
+            for student_id, submission_files in submissions.items():
+                for text in submission_files['texts']:
+                    text_path = os.path.join(
+                        str(course_id),
+                        block_id,
+                        student_id,
+                        text['filename'],
+                    )
+
+                    zip_file.writestr(text_path, text['text'])
+
+                for attachment in submission_files['attachments']:
+                    attachment_path = os.path.join(
+                        str(course_id),
+                        block_id,
+                        student_id,
+                        'attachments',
+                        attachment['name'],
+                    )
+
+                    with storage.open(attachment['storage_path']) as attachment_file:
+                        zip_file.writestr(attachment_path, attachment_file.read())
+
+        downloads_csv = _create_downloads_csv(submission_files_data)
+        downloads_csv_path = os.path.join(str(course_id), 'downloads.csv')
+
+        zip_file.writestr(
+            downloads_csv_path,
+            downloads_csv.getvalue().encode('utf-8')
+        )
+
+    file.seek(0)
+
+    return True
+
+
+def upload_ora2_submission_files(
+    _xmodule_instance_args, _entry_id, course_id, _task_input, action_name
+):
+    """
+    Creates zip archive with submission files in three steps:
+
+    1. Collect all files information using ORA aggregation helper.
+    2. Download all submission attachments, put them in temporary zip file along
+        with submission texts and csv downloads list.
+    3. Upload zip file into reports storage.
+    """
+
+    start_time = time()
+    start_date = datetime.now(UTC)
+
+    num_attempted = 1
+    num_total = 1
+
+    fmt = 'Task: {task_id}, InstructorTask ID: {entry_id}, Course: {course_id}, Input: {task_input}'
+    task_info_string = fmt.format(
+        task_id=_xmodule_instance_args.get('task_id') if _xmodule_instance_args is not None else None,
+        entry_id=_entry_id,
+        course_id=course_id,
+        task_input=_task_input
+    )
+    TASK_LOG.info(u'%s, Task type: %s, Starting task execution', task_info_string, action_name)
+
+    task_progress = TaskProgress(action_name, num_total, start_time)
+    task_progress.attempted = num_attempted
+
+    submission_files_data = _task_step(
+        lambda: OraAggregateData.collect_ora2_submission_files(course_id),
+        task_progress,
+        task_info_string,
+        action_name,
+        'Collecting attachments data',
+        'Failed to get ORA submissions attachments data.',
+        'Error while collecting data',
+    )
+    if submission_files_data is None:
+        return UPDATE_STATUS_FAILED
```

Comment:
If _task_step is made into a context manager, this becomes:

```python
    with _task_step(
        task_progress,
        task_info_string,
        action_name,
        'Collecting attachments data',
        'Failed to get ORA submissions attachments data.',
        'Error while collecting data',
    ):
        submission_files_data = OraAggregateData.collect_ora2_submission_files(course_id)
        if submission_files_data is None:
            return UPDATE_STATUS_FAILED
```

In fact, since the first three parameters are passed multiple times, you can make _task_step take the first 3 params and return a custom context manager with that data.